src/stan/services/sample/hmc_nuts_diag_e_adapt.hpp

```diff
@@ -143,17 +147,219 @@ int hmc_nuts_diag_e_adapt(
+    callbacks::interrupt& interrupt, callbacks::logger& logger,
+    std::vector<InitWriter>& init_writer,
+    std::vector<SampleWriter>& sample_writer,
+    std::vector<DiagnosticWriter>& diagnostic_writer) {
+  if (num_chains == 1) {
+    return hmc_nuts_diag_e_adapt(
+        model, *init[0], *init_inv_metric[0], random_seed, init_chain_id,
+        init_radius, num_warmup, num_samples, num_thin, save_warmup, refresh,
+        stepsize, stepsize_jitter, max_depth, delta, gamma, kappa, t0,
+        init_buffer, term_buffer, window, interrupt, logger, init_writer[0],
+        sample_writer[0], diagnostic_writer[0]);
+  } else {
+    using sample_t = stan::mcmc::adapt_diag_e_nuts<Model, boost::ecuyer1988>;
+    std::vector<boost::ecuyer1988> rngs;
+    rngs.reserve(num_chains);
+    std::vector<std::vector<double>> cont_vectors;
+    cont_vectors.reserve(num_chains);
+    std::vector<sample_t> samplers;
+    samplers.reserve(num_chains);
+    try {
+      for (int i = 0; i < num_chains; ++i) {
+        rngs.emplace_back(util::create_rng(random_seed, init_chain_id, i));
+        cont_vectors.emplace_back(util::initialize(model, *init[i], rngs[i],
+                                                   init_radius, true, logger,
+                                                   init_writer[i]));
+        samplers.emplace_back(model, rngs[i]);
+        Eigen::VectorXd inv_metric = util::read_diag_inv_metric(
+            *init_inv_metric[i], model.num_params_r(), logger);
+        util::validate_diag_inv_metric(inv_metric, logger);
+
+        samplers[i].set_metric(inv_metric);
+        samplers[i].set_nominal_stepsize(stepsize);
+        samplers[i].set_stepsize_jitter(stepsize_jitter);
+        samplers[i].set_max_depth(max_depth);
+
+        samplers[i].get_stepsize_adaptation().set_mu(log(10 * stepsize));
+        samplers[i].get_stepsize_adaptation().set_delta(delta);
+        samplers[i].get_stepsize_adaptation().set_gamma(gamma);
+        samplers[i].get_stepsize_adaptation().set_kappa(kappa);
+        samplers[i].get_stepsize_adaptation().set_t0(t0);
+        samplers[i].set_window_params(num_warmup, init_buffer, term_buffer,
+                                      window, logger);
+      }
+    } catch (const std::domain_error& e) {
+      return error_codes::CONFIG;
+    }
+    tbb::parallel_for(
+        tbb::blocked_range<size_t>(0, num_chains, 1),
+        [num_warmup, num_samples, num_thin, refresh, save_warmup, num_chains,
+         &samplers, &model, &rngs, &interrupt, &logger, &sample_writer,
+         &cont_vectors,
+         &diagnostic_writer](const tbb::blocked_range<size_t>& r) {
+          for (size_t i = r.begin(); i != r.end(); ++i) {
+            util::run_adaptive_sampler(
+                samplers[i], model, cont_vectors[i], num_warmup, num_samples,
+                num_thin, refresh, save_warmup, rngs[i], interrupt, logger,
+                sample_writer[i], diagnostic_writer[i], i + 1, num_chains);
```

Comment:
the chain-id given to `run_adaptive_sampler` is `i+1`... which labels from 1 to num_chains... but shouldn't we make this `init_chain_id + i`? I do think so. In case you like 1...num_chains  numbering then the least we need to do is to aling the argument name for the downstream functions. That is right now called `chain_id`... which it is not if we don't change it as suggested.

I'd rather like to get the actual chain id's being displayed.

Reply:
I think what we want is `init_chain_id + i + 1` right? for `n_chain = 1` `generate_transitions` gives the output we have now, but for `n_chain > 1` I think we want the initial chain id + what chain we are on. So for `init_chain_id = 5` and `n_chain = 4` we would get back 

```bash
Chain [5] Iteration: ... 
Chain [6] Iteration: ... 
Chain [7] Iteration: ... 
Chain [8] Iteration: ... 
```

I'd prefer not changing the output for the default/current stuff. I'm sure someone somewhere is logging and grepping that and I'd rather not cause a headache somewhere

Reply:
I think it's `init_chain_id + i`, because `init_chain_id` is the first chain id you want to see.

I am ok with leaving the old code around and have slightly different outputs for multi chain vs single chain runs.

This point still needs update in the code, right?

Reply:
> I think it's init_chain_id + i, because init_chain_id is the first chain id you want to see.

Do we want to index from 0 or 1 for the printing? Looking at the code for `generate_transitions()` if we did `init_chain_id + i` for 4 chains and init value of 0 (the default for cmdstan) we will have 

```
Chain [0] Iteration: ... 
Chain [1] Iteration: ... 
Chain [2] Iteration: ... 
Chain [3] Iteration: ... 
```

> This point still needs update in the code, right?

If we are indexing from 1 then the code I have right now is right, but if want to index from zero then I need to remove the `+ 1`

Reply:
If we index from 0 right now, then we should display it like that, I think. I would find it very confusing to have the printed id to be different than the actual id of the chain.

@mitzimorris what do you think?

Reply:
Sorry, by

> If we index from 0 right now, then we should display it like that, I think

what indexing are you referring to? The initial id? That is a value of zero. idk if there is a current precedent for this. But if we are going to do the chain outputs as `{output_name}_{chain_id}.csv` starting from zero then I do think that it makes sense to index from 0 here. Does that make sense? I think what I'm saying is that if the output csv's names start from 0 we should have these start from 0 as well. Though we could change those to start at 1. I'm fine either way, though I kind of like indexing from 1 since the Stan language indexes from 1.

Reply:
OH... good point. So yeah, the outputs should be aligned with the chain id is my view here. It is unfortunate that cmdstan allows chain ids to start with 0 and even the default is 0. My proposal is to align chain id with whatever number we show on screen or put into file names. I would find it extremely confusing to see chain id 2 being sampled in the initial print and then the chain is being referred to as chain 3??

In order to remedy the situation we can consider to change the default chain id in cmdstan to be 1 instead of 0 (while still allowing a chain id of 0).

Reply:
Yep I'm cool with that! Just made the change to `init_chain_id + i`. We can talk about changing the default init id to 0 in the cmdstan branch